crates/sui-replay-2/src/data_store.rs

```diff
@@ -0,0 +1,428 @@
+            ReplayError::TransactionEffectsConversionError {
+                id: tx_digest.to_string(),
+                err: format!("{:?}", e),
+            }
+        })?;
+        debug!("End transaction effects");
+        Ok(effects)
+    }
+
+    //
+    // Epoch table eager load
+    //
+    async fn epochs_gql_table(client: &Client) -> Result<BTreeMap<u64, EpochData>, ReplayError> {
+        debug!("Start load epoch table");
+        let mut pag_filter = PaginationFilter {
+            direction: Direction::Backward,
+            cursor: None,
+            limit: None,
+        };
+
+        let mut epochs_data = BTreeMap::<u64, EpochData>::new();
+        loop {
+            debug!("Start load epoch table pagination");
+            let paged_epochs = crate::gql_queries::epochs(client, pag_filter)
+                .await
+                .map_err(|e| {
+                    let err = format!("{:?}", e);
+                    ReplayError::GenericError { err }
+                })
+                .unwrap();
+            let (page_info, data) = paged_epochs.into_parts();
+            for epoch in data {
+                epochs_data.insert(epoch.epoch_id, epoch.try_into()?);
+            }
+            if page_info.has_previous_page {
+                pag_filter = PaginationFilter {
+                    direction: Direction::Backward,
+                    cursor: page_info.start_cursor,
+                    limit: None,
+                };
+            } else {
+                break;
+            }
+        }
+
+        debug!("End load epoch table");
+        Ok(epochs_data)
+    }
+
+    //
+    // Dynamic fields operations
+    //
+
+    pub fn read_child_object(
+        &self,
+        _parent: &ObjectID,
+        child: &ObjectID,
+        child_version_upper_bound: SequenceNumber,
+    ) -> Result<Option<Object>, ReplayError> {
+        #[allow(clippy::disallowed_methods)]
+        let obj = futures::executor::block_on(dynamic_field_at_version(
+            &self.client,
+            *child,
+            child_version_upper_bound.value(),
+        ))
+        .map_err(|e| ReplayError::DynamicFieldError { err: e.to_string() })?;
+        match obj {
+            None => Ok(None),
+            Some(obj) => {
+                let obj = Object::try_from(obj)
+                    .map_err(|e| ReplayError::DynamicFieldError { err: e.to_string() })?;
+                Ok(Some(obj))
+            }
+        }
+    }
+}
+
+type EpochId = u64;
+
+// Eager loading of the epoch table from GQL.
+// Maps an epoch to data vital to trascation execution:
+// framework versions, protocol version, RGP, epoch start timestamp
+#[derive(Debug)]
+pub struct EpochStore {
+    data: BTreeMap<EpochId, EpochData>,
+}
+
+impl EpochStore {
+    fn eager(data: BTreeMap<EpochId, EpochData>) -> Self {
+        Self { data }
+    }
+
+    fn lazy() -> Self {
+        Self {
+            data: BTreeMap::new(),
```

Comment:
I'm guessing `lazy ` will do something more interesting later? If so it may be worthwhile marking this with a TODO.